File: aoc2020/src/day6/part1.ts
async function main() {
  const file = Bun.file("./input.txt");
  const arrBuffer = await file.arrayBuffer();
  const buffer = Buffer.from(arrBuffer);
  const data = buffer.toString();

  const splitGroups = (input: string): string[][] => {
    return input
      .trim()
      .split("\n\n")
      .map((grp) => grp.split("\n"));
  };

  const countUniqueYes = (group: string[]): number => {
    const allAns = new Set(group.join("").split(""));
    return allAns.size;
  };

  const countTotalUnique = (input: string): number => {
    const groups = splitGroups(input);
    return groups.reduce((total, group) => total + countUniqueYes(group), 0);
  };

  console.log(countTotalUnique(data));
}

await main();

File: aoc2020/day6/part1.py
def split_groups(input_str: str) -> list[list[str]]:
    return [group.split("\n") for group in input_str.strip().split("\n\n")]

def count_unique_yes(group: list[str]) -> int:
    # Join all answers in group and convert to set of individual characters
    all_answers = set(''.join(group))
    return len(all_answers)

def count_total_unique(input_str: str) -> int:
    groups = split_groups(input_str)
    return sum(count_unique_yes(group) for group in groups)

def main():
    with open("input.txt", "r") as file:
        data = file.read()
    print(count_total_unique(data))

if __name__ == "__main__":
    main()
